Enforce Quality Gates for Java and Python sub projects on CI

ndegwamartin opened the pull request:
## Description of what I changed

Resolves #1541 

## E2E test

<!-- There are different scenarios for using the tools in this repo; please
  help your reviewers by describing how you have e2e tested your change. -->

TESTED:

Please replace this with a description of how you tested your PR beyond the
automated e2e/unit tests.

## Checklist: I completed these to help reviewers :)

<!--- Put an `x` in the box if you did the task -->
<!--- If you forgot a task please follow the instructions below -->

- [ ] I have read and will follow the
      [review process](https://github.com/GoogleCloudPlatform/openmrs-fhir-analytics/blob/master/doc/review_process.md).
- [ ] I am familiar with Google Style Guides for the language I have coded in.

  No? Please take some time and review
  [Java](https://google.github.io/styleguide/javaguide.html) and
  [Python](https://google.github.io/styleguide/pyguide.html) style guides.

- [ ] My IDE is configured to follow the Google
      [**code styles**](https://google.github.io/styleguide/).

  No? Unsure? ->
  [configure your IDE](https://github.com/google/google-java-format).

- [ ] I have **added tests** to cover my changes. (If you refactored existing
      code that was well tested you do not have to add tests)
- [ ] I ran `mvn clean package` right before creating this pull request and
      added all formatting changes to my commit.
- [ ] If I made any Python code changes, I ran `black .`, `pylint .` and
      `pyright` . right before creating this pull request and added all
      formatting changes to my commit.
- [ ] All new and existing **tests passed**.
- [ ] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

ndegwamartin committed:
Fixes ErrorProne warnings on master

ndegwamartin committed:
Enables CI Gate check for Java checks

ndegwamartin committed:
Updates Maven configuration for ErrorProne

ndegwamartin committed:
Sets up Github actions workflow for python code quality

ndegwamartin committed:
Refactors f-string logger interpolations to use lazy % format, fixes pylint error W1203

ndegwamartin committed:
Removes Python static checks for synthea sub-projects

ndegwamartin committed:
Fixes Pylint error W1405: Inconsistent-quotes

ndegwamartin committed:
Updates Python Code Quality workflow configuration

ndegwamartin committed:
Updates Python lint workflow to include pip dev dependencies

bashir2 commented on bunsen/bunsen-avro/src/main/java/com/cerner/bunsen/avro/converters/DefinitionToAvroVisitor.java:
```diff
@@ -37,7 +37,6 @@
 import org.apache.avro.JsonProperties;
 import org.apache.avro.Schema;
 import org.apache.avro.Schema.Field;
-import org.apache.avro.Schema.Type;
```
Are these `Schema.Type` related changes required? If not, can we revert these to make the PR more focused?

bashir2 replied:
Okay, it would be nice to do that for `Schema.Field` too for consistency, but up to you.

ndegwamartin replied:
Done

bashir2 commented on .github/workflows/python-lint.yml:
```diff
@@ -0,0 +1,46 @@
+name: Python Code Quality
```
If these workflows fail, I suppose it is easy for the developer to reproduce the failure locally (using the instructions you have added recently); correct?

ndegwamartin replied:
Yeah so basically this workflow here automates enforcing that the developer actually ran the python code quality tools using our instructions on the Pull Request Template. i.e. reproducing entails just running the checks locally which is already documented e.g. for formatting: 

```bash
$ black .
```

bashir2 reviewed (commented):
Thanks @ndegwamartin for the changes; just a few minor suggestions/questions.

ndegwamartin commented on pipelines/batch/src/main/java/com/google/fhir/analytics/HapiRowDescriptor.java:
```diff
@@ -32,7 +32,7 @@ abstract class HapiRowDescriptor implements Serializable {
   static HapiRowDescriptor create(
       String resourceId,
-      String forcedId,
+      @Nullable String forcedId,
```
I think the tool is flagging it because of this annotation here which @Autovalue respects when it generates the class constructor:
https://github.com/google/fhir-data-pipes/blob/master/pipelines/batch/src/main/java/com/google/fhir/analytics/HapiRowDescriptor.java#L54

bashir2 commented on pipelines/controller/src/main/java/com/google/fhir/analytics/PipelineManager.java:
```diff
@@ -174,7 +175,7 @@ public CumulativeMetrics getCumulativeMetrics() {
   private void setLastRunStatus(LastRunStatus status) {
     if (status == LastRunStatus.SUCCESS) {
-      lastRunEnd = LocalDateTime.now();
+      lastRunEnd = LocalDateTime.now(ZoneOffset.UTC);
```
So as discussed please either add a unit-test for checking the incremental runs are triggered properly or do a manual test with your changes included.

ndegwamartin replied:
Added a static helper function in `DwhFilesManager` and used in both classes. Let me know if a stand alone Utils class is preferred.

bashir2 reviewed (approved):
Thanks @ndegwamartin please feel free to merge after addressing the remaining comments below.

ndegwamartin committed:
Address PR Feedback
- Reverts lazy logging for main.py
- Disables lazy logging check in pylint configuration
- Refactors get current time to use static helper method
- Adds unit tests for PipelineManager class

ndegwamartin committed:
Pin python dependencies versions
-Fix pylint warning

ndegwamartin committed:
Clean up

bashir2 commented on pipelines/controller/src/test/java/com/google/fhir/analytics/PipelineManagerTest.java:
```diff
@@ -0,0 +1,129 @@
+          InvocationTargetException.class,
+          () -> {
+            // Call checkSchedule via reflection
+            Method checkScheduleMethod = PipelineManager.class.getDeclaredMethod("checkSchedule");
```
Instead of using reflection, please use `@VisibleForTesting` annotation, for example like [here](https://github.com/google/fhir-data-pipes/blob/547612dea431ebec8800fb99f9674150e06dce57/pipelines/common/src/main/java/com/google/fhir/analytics/AvroConversionUtil.java#L154). You can do it both for methods and attributes (so can avoid other reflection uses as well). But please also try to organize the test to minimize such annotations.

ndegwamartin replied:
I've also refactored the logic for the `testIncrementalModeTriggeredAtRightTime` test case for completeness.

ndegwamartin committed:
Address PR Feedback II

bashir2 reviewed (approved):
Just some minor unit-test/architecture related suggestion.

ndegwamartin committed:
Refactors PipelineManager to constructor injection for testability

ndegwamartin committed:
Merge branch 'master' into enable-ci-gates

bashir2 commented on pipelines/controller/src/test/java/com/google/fhir/analytics/PipelineManagerTest.java:
```diff
@@ -21,35 +21,33 @@
-    setField("dwhFilesManager", dwhFilesManager);
-    setField("meterRegistry", meterRegistry);
+    pipelineManager = new PipelineManager(dataProperties, dwhFilesManager, meterRegistry);
     setField("cron", CronExpression.parse("0 * * * * *")); // every minute
```
Instead of using injection, can we tweak the test `dataProperties` above to return the required `incrementalSchedule`?

ndegwamartin replied:
Done.

ndegwamartin replied:
resolved

bashir2 reviewed (approved):
Minor

ndegwamartin committed:
Merge branch 'master' into enable-ci-gates

ndegwamartin committed:
Addresses QA feedback

bashir2 reviewed (commented):
Please feel free to merge after addressing minor test comments.

ndegwamartin committed:
Address PR feedback

ndegwamartin committed:
Merge branch 'master' into enable-ci-gates

ndegwamartin committed:
Trigger CI

ndegwamartin committed:
Trigger CI

ndegwamartin committed:
Fixes failing unit test
